File: MusicSalesApp/Components/Shared/LyricsScroller.razor.js
/*
 * Karaoke highlighting and scrolling.
 *
 * WHY THIS IS HERE AND NOT IN C#
 * ------------------------------
 * Every player in this application already reports its position to .NET through the audio element's
 * `timeupdate` event. That is the obvious source for a highlighter and it cannot be used for one:
 * `timeupdate` fires roughly four times a second, and this is Blazor SERVER, so each tick is a
 * SignalR round trip ending in StateHasChanged over the whole page. Highlighting a word four times a
 * second, late by a network hop, on a page that might have a dozen playing cards, is both visibly
 * wrong and expensive.
 *
 * So the words are rendered once by Blazor with data-l/data-w attributes, and everything that has to
 * happen per frame happens here, against the audio element's own clock. Nothing in this file calls
 * back into .NET except a word click, which happens at human speed.
 *
 * ONE MAP, NOT MODULE-LEVEL STATE
 * -------------------------------
 * The library grid renders one scroller per card and Blazor loads this module once for all of them,
 * so every instance's state is keyed by an id minted in C#.
 */

const instances = new Map();

/** Below this the line is considered already centred; avoids a transform on every frame. */
const RECENTRE_EPSILON_PX = 1;

export function init(id, root, audio, dotNetRef, interactive) {
    // Checked by capability, not by truthiness. An ElementReference that was never assigned - which
    // happens whenever the audio element is rendered after this component in its parent's markup -
    // arrives here as an object rather than as null, so `!audio` does not catch it. Calling
    // addEventListener on it throws, and an unhandled JSException on Blazor Server does not fail
    // quietly: it tears down the circuit and the whole page reports "An unhandled error has occurred".
    if (!root || !audio || typeof audio.addEventListener !== 'function') {
        return;
    }

    dispose(id);

    const state = {
        root,
        audio,
        dotNetRef,
        track: root.querySelector('.lyrics-scroller-track'),
        words: [],
        lines: [],
        starts: [],
        ends: [],
        lineOf: [],
        wordOf: [],
        activeIndex: -1,
        activeLine: -1,
        frame: 0,
        listeners: [],
    };

    const on = (target, event, handler) => {
        target.addEventListener(event, handler);
        state.listeners.push([target, event, handler]);
    };

    // The loop runs only while the audio is actually playing. An idle 60fps loop per card is pure
    // waste on a page showing a grid of them.
    on(audio, 'play', () => start(state));
    on(audio, 'playing', () => start(state));
    on(audio, 'pause', () => { stop(state); update(state); });
    on(audio, 'ended', () => { stop(state); update(state); });

    // Seeking has to update even while paused - a creator scrubbing to find a line expects the
    // highlight to follow, and the rAF loop is not running to do it for them.
    on(audio, 'seeked', () => update(state, true));
    on(audio, 'seeking', () => update(state, true));
    on(audio, 'loadedmetadata', () => update(state, true));

    if (interactive) {
        // ONE delegated listener, not one per word. A four-minute song is a few thousand words.
        on(root, 'click', (event) => {
            const target = event.target.closest('[data-w]');
            if (!target || !state.dotNetRef) {
                return;
            }

            const line = Number(target.getAttribute('data-l'));
            const word = Number(target.getAttribute('data-w'));

            if (Number.isInteger(line) && Number.isInteger(word)) {
                state.dotNetRef.invokeMethodAsync('WordClicked', line, word);
            }
        });
    }

    instances.set(id, state);
    cacheElements(state);
}

/** Hand over timings the caller already had (the editor, and any draft). */
export function setTimings(id, timings) {
    const state = instances.get(id);
    if (!state || !timings) {
        return;
    }

    state.starts = timings.starts || [];
    state.ends = timings.ends || [];
    state.lineOf = timings.lineOf || [];
    state.wordOf = timings.wordOf || [];
    state.activeIndex = -1;
    state.activeLine = -1;

    cacheElements(state);
    update(state, true);

    if (!state.audio.paused) {
        start(state);
    }
}

/** Fetch published timings over the public media route. */
export async function loadTimings(id, url) {
    const state = instances.get(id);
    if (!state || !url) {
        return;
    }

    try {
        const response = await fetch(url, { credentials: 'same-origin' });
        if (!response.ok) {
            // A 404 here is the normal answer for timings the creator has not published, so this is
            // not an error condition - the scroller simply has nothing to show.
            return;
        }

        const document = await response.json();
        setTimings(id, flatten(document));
    } catch {
        // Network trouble costs the lyrics, not the page.
    }
}

/**
 * Flatten the document the same way the C# side does.
 *
 * Only used on the fetch path: the published JSON is the Python writer's nested shape, and the
 * per-frame lookup wants one ascending array rather than a tree to descend.
 */
function flatten(document) {
    const starts = [];
    const ends = [];
    const lineOf = [];
    const wordOf = [];

    (document.lines || []).forEach((line, lineIndex) => {
        if (line.startMs === null || line.startMs === undefined) {
            return;
        }

        const words = line.words || [];

        if (words.length === 0) {
            starts.push(line.startMs);
            ends.push(line.endMs);
            lineOf.push(lineIndex);
            wordOf.push(0);
            return;
        }

        words.forEach((word, wordIndex) => {
            if (word.startMs === null || word.startMs === undefined) {
                return;
            }
            starts.push(word.startMs);
            ends.push(word.endMs);
            lineOf.push(lineIndex);
            wordOf.push(wordIndex);
        });
    });

    return { starts, ends, lineOf, wordOf, durationMs: document.durationMs };
}

/**
 * Re-read the DOM after Blazor has rendered a different document.
 *
 * Cached because querySelectorAll on every frame, for every playing card, is exactly the kind of
 * cost that makes a page feel heavy for no visible reason.
 */
function cacheElements(state) {
    state.words = Array.from(state.root.querySelectorAll('[data-w]'));
    state.lines = Array.from(state.root.querySelectorAll('.lyrics-line[data-l]'));
    state.track = state.root.querySelector('.lyrics-scroller-track');

    state.wordByKey = new Map();
    state.words.forEach((element) => {
        state.wordByKey.set(`${element.getAttribute('data-l')}:${element.getAttribute('data-w')}`, element);
    });

    state.lineByIndex = new Map();
    state.lines.forEach((element) => {
        state.lineByIndex.set(Number(element.getAttribute('data-l')), element);
    });
}

function start(state) {
    if (state.frame) {
        return;
    }

    const tick = () => {
        update(state);
        state.frame = requestAnimationFrame(tick);
    };

    state.frame = requestAnimationFrame(tick);
}

function stop(state) {
    if (state.frame) {
        cancelAnimationFrame(state.frame);
        state.frame = 0;
    }
}

/**
 * Move the highlight to wherever the audio actually is.
 *
 * The position is re-derived from currentTime every frame rather than accumulated locally, so
 * buffering, a seek, or a tab that was backgrounded all correct themselves on the next frame instead
 * of drifting permanently.
 */
function update(state, force) {
    const count = state.starts.length;
    if (count === 0) {
        return;
    }

    const ms = state.audio.currentTime * 1000;
    let index = state.activeIndex;

    if (force || index < 0 || index >= count || ms < state.starts[index]) {
        index = search(state.starts, ms);
    } else {
        // The common case: time moved forward a frame's worth, so the answer is here or just after.
        while (index + 1 < count && state.starts[index + 1] <= ms) {
            index++;
        }
    }

    if (index === state.activeIndex && !force) {
        return;
    }

    state.activeIndex = index;

    const previous = state.root.querySelector('.lyrics-word--active');
    if (previous) {
        previous.classList.remove('lyrics-word--active');
    }

    if (index < 0) {
        return;
    }

    const element = state.wordByKey.get(`${state.lineOf[index]}:${state.wordOf[index]}`);
    if (element) {
        element.classList.add('lyrics-word--active');
    }

    const line = state.lineOf[index];
    if (line !== state.activeLine) {
        state.activeLine = line;
        highlightLine(state, line);
        centre(state, line);
    }
}

/** The last index whose start is at or before `ms`, or -1 before the first word. */
function search(starts, ms) {
    let low = 0;
    let high = starts.length - 1;
    let found = -1;

    while (low <= high) {
        const mid = (low + high) >> 1;
        if (starts[mid] <= ms) {
            found = mid;
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    return found;
}

function highlightLine(state, lineIndex) {
    const previous = state.root.querySelector('.lyrics-line--active');
    if (previous) {
        previous.classList.remove('lyrics-line--active');
    }

    const element = state.lineByIndex.get(lineIndex);
    if (element) {
        element.classList.add('lyrics-line--active');
    }
}

/**
 * Bring the current line to the middle by TRANSLATING THE TRACK, never by scrolling.
 *
 * scrollIntoView({ block: 'center' }) would be one line of code and would scroll every scrollable
 * ancestor to get there - yanking the library grid, or the player page, or a dialog, whenever a song
 * moved to the next line. Moving a child inside a fixed, clipped box cannot affect anything outside
 * it. The easing is a CSS transition on the track, so this stays one style write per line change.
 */
function centre(state, lineIndex) {
    const element = state.lineByIndex.get(lineIndex);
    if (!element || !state.track) {
        return;
    }

    const target = element.offsetTop + (element.offsetHeight / 2) - (state.root.clientHeight / 2);
    const next = -Math.max(0, target);
    const current = state.currentOffset ?? 0;

    if (Math.abs(next - current) < RECENTRE_EPSILON_PX) {
        return;
    }

    state.currentOffset = next;
    state.track.style.transform = `translate3d(0, ${next}px, 0)`;
}

/** Mark a word as selected in the editor. Purely visual; selection itself lives in C#. */
export function setSelectedWord(id, lineIndex, wordIndex) {
    const state = instances.get(id);
    if (!state) {
        return;
    }

    const previous = state.root.querySelector('.lyrics-word--selected');
    if (previous) {
        previous.classList.remove('lyrics-word--selected');
    }

    const element = state.wordByKey.get(`${lineIndex}:${wordIndex}`);
    if (element) {
        element.classList.add('lyrics-word--selected');
    }
}

export function dispose(id) {
    const state = instances.get(id);
    if (!state) {
        return;
    }

    stop(state);

    state.listeners.forEach(([target, event, handler]) => {
        target.removeEventListener(event, handler);
    });

    state.listeners = [];
    state.words = [];
    state.lines = [];
    state.wordByKey = null;
    state.lineByIndex = null;

    instances.delete(id);
}

File: MusicSalesApp/Components/Shared/LyricsScroller.razor.cs
#nullable enable
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using MusicSalesApp.Common.Contracts;

namespace MusicSalesApp.Components.Shared;

/// <summary>How much room the scroller has, which is the only thing that varies between its homes.</summary>
public enum LyricsScrollerMode
{
    /// <summary>A song card in the library grid. Smallest type, fewest lines visible.</summary>
    Card,

    /// <summary>The song or playlist player page, in place of the cover art.</summary>
    Player,

    /// <summary>The creator's timing editor. Same rendering, plus word selection.</summary>
    Editor
}

/// <summary>
/// Scrolling lyrics with the sung word highlighted, in whichever of its three homes.
///
/// <para>
/// <b>The highlighting does not run in C#, and that is the central decision here.</b> The obvious
/// implementation drives it from the existing <c>timeupdate</c> callback, which is what every player
/// on this site already uses to move its progress bar. It cannot work for this: <c>timeupdate</c>
/// fires about four times a second, and on Blazor <em>Server</em> every one of those is a SignalR
/// round trip that ends in <c>StateHasChanged</c> over the whole page. Karaoke highlighting at 4 Hz
/// is visibly late and jerky, and paying a render diff per tick - per playing card, on a page that
/// can show a dozen - is a cost nobody would accept once they saw it.
/// </para>
///
/// <para>
/// So the component renders the words once, with <c>data-l</c>/<c>data-w</c> attributes, and hands
/// the timings to a JS module that runs a <c>requestAnimationFrame</c> loop against the audio
/// element's own clock. Blazor is not in the hot path at all. What C# keeps is everything that
/// happens at human speed: which document is loaded, which word is selected, and teardown.
/// </para>
/// </summary>
public class LyricsScrollerModel : ComponentBase, IAsyncDisposable
{
    [Inject] private IJSRuntime JS { get; set; } = default!;
    [Inject] private ILogger<LyricsScrollerModel> Logger { get; set; } = default!;

    /// <summary>The audio element these lyrics follow. Supplied by whichever player owns it.</summary>
    [Parameter, EditorRequired] public ElementReference AudioElement { get; set; }

    /// <summary>How much room this instance has.</summary>
    [Parameter] public LyricsScrollerMode Mode { get; set; } = LyricsScrollerMode.Player;

    /// <summary>
    /// The timings to display, when the caller already has them - the editor, which reads drafts
    /// through the service because unpublished timings are not fetchable over HTTP.
    /// </summary>
    [Parameter] public LyricsTimingsDocument? Document { get; set; }

    /// <summary>
    /// Where to fetch published timings instead, relative to the site root.
    ///
    /// <para>
    /// Must carry the <c>?v=</c> cache-buster. The blob path never changes between versions and the
    /// response is served with a year-long immutable cache header, so without it a creator's
    /// re-publish is invisible to every browser that has already seen the song - permanently.
    /// </para>
    /// </summary>
    [Parameter] public string? TimingsUrl { get; set; }

    /// <summary>Raised when a word is clicked, in editor mode only.</summary>
    [Parameter] public EventCallback<LyricsWordSelection> OnWordSelected { get; set; }

    /// <summary>What to show before any timings arrive.</summary>
    [Parameter] public string EmptyMessage { get; set; } = "No lyrics yet.";

    /// <summary>
    /// Identifies this instance to the JS module.
    ///
    /// <para>
    /// Needed because the library grid shares one module across every card on the page, so the module
    /// keeps its loops in a map rather than in module-level variables.
    /// </para>
    /// </summary>
    protected string InstanceId { get; } = $"ls-{Guid.NewGuid():N}";

    protected ElementReference _root;
    protected LyricsTimingsDocument? _document;

    private IJSObjectReference? _module;
    private DotNetObjectReference<LyricsScrollerModel>? _selfRef;
    private bool _initialised;
    private string? _pushedSignature;

    protected override void OnParametersSet()
    {
        // Rendered from whichever source the caller gave. A fetched document arrives later, through
        // JS, and never round-trips into C# - there is nothing here that needs to see it.
        if (Document is not null)
        {
            _document = Document;
        }
    }

    /// <summary>
    /// Initialise once the audio element actually exists, which is not necessarily the first render.
    ///
    /// <para>
    /// <b>Deliberately not gated on <c>firstRender</c>.</b> <see cref="ElementReference"/> is a struct
    /// captured when the parameter is set, and a parent that declares its <c>&lt;audio&gt;</c> after
    /// this component - as the timing editor does - has not assigned its <c>@ref</c> yet at that
    /// point, because refs are populated only once the render batch reaches the DOM. Initialising
    /// anyway hands JS a reference to nothing, and the resulting JSException takes the circuit down
    /// with it. Waiting for a populated <see cref="ElementReference.Id"/> costs one extra render and
    /// removes the ordering trap entirely.
    /// </para>
    /// </summary>
    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (!_initialised)
        {
            if (string.IsNullOrEmpty(AudioElement.Id))
            {
                return;
            }

            _selfRef ??= DotNetObjectReference.Create(this);

            try
            {
                _module ??= await JS.InvokeAsync<IJSObjectReference>(
                    "import", "./Components/Shared/LyricsScroller.razor.js");

                await _module.InvokeVoidAsync(
                    "init",
                    InstanceId,
                    _root,
                    AudioElement,
                    _selfRef,
                    Mode == LyricsScrollerMode.Editor);

                _initialised = true;
            }
            catch (JSDisconnectedException)
            {
                // The circuit went away mid-initialisation. Nothing to clean up that dispose will not
                // handle, and nothing worth logging as an error.
                return;
            }
        }

        if (_initialised)
        {
            await PushTimingsAsync();
        }
    }

    /// <summary>
    /// Hand the module whatever it should be following now.
    ///
    /// <para>
    /// Guarded by a signature so a re-render caused by something unrelated - a parent's state change,
    /// a toggle - does not re-send a document the module already has and restart its loop mid-song.
    /// </para>
    /// </summary>
    private async Task PushTimingsAsync()
    {
        var signature = Document is not null
            ? $"doc:{Document.SongId}:{Document.DurationMs}:{Document.Lines.Count}:{DocumentStamp(Document)}"
            : $"url:{TimingsUrl}";

        if (signature == _pushedSignature || _module is null)
        {
            return;
        }

        _pushedSignature = signature;

        try
        {
            if (Document is not null)
            {
                await _module.InvokeVoidAsync("setTimings", InstanceId, Flatten(Document));
            }
            else if (!string.IsNullOrWhiteSpace(TimingsUrl))
            {
                await _module.InvokeVoidAsync("loadTimings", InstanceId, TimingsUrl);
            }
        }
        catch (JSDisconnectedException)
        {
            // Circuit gone. Nothing to do.
        }
        catch (Exception ex)
        {
            Logger.LogWarning(ex, "Could not hand lyric timings to the scroller.");
        }
    }

    /// <summary>
    /// A cheap fingerprint of the document's timings, so an edit is noticed but a re-render is not.
    /// </summary>
    private static long DocumentStamp(LyricsTimingsDocument document)
    {
        long stamp = 17;

        foreach (var line in document.Lines)
        {
            stamp = unchecked(stamp * 31 + (line.StartMs ?? -1));
            stamp = unchecked(stamp * 31 + (line.EndMs ?? -1));
        }

        return stamp;
    }

    /// <summary>
    /// The line's span as <c>2:45.5 - 2:45.6</c>, shown down the editor's left edge.
    ///
    /// <para>
    /// Tenths rather than hundredths on purpose: a creator is comparing these against what they can
    /// hear, and hundredths are past the resolution of that judgement. Tenths are also enough to make
    /// the failure this was added for obvious - a line squeezed into a few hundredths reads as
    /// <c>2:45.5 - 2:45.6</c>, where the highlight simply appears never to land on it.
    /// </para>
    /// </summary>
    protected static string FormatLineRange(LyricsTimedLine line) =>
        line.IsTimed
            ? $"{FormatStamp(line.StartMs!.Value)} - {FormatStamp(line.EndMs!.Value)}"
            : string.Empty;

    private static string FormatStamp(long ms)
    {
        if (ms < 0)
        {
            ms = 0;
        }

        var minutes = ms / 60_000;
        var seconds = (ms % 60_000) / 1000d;

        return $"{minutes}:{seconds:00.0}";
    }

    /// <summary>
    /// Flatten the document into parallel arrays of every <em>timed</em> word.
    ///
    /// <para>
    /// The module does a lookup on every animation frame, so what it wants is one ascending array of
    /// start times it can binary-search or walk, not a tree it has to descend. Sending the nested
    /// document and flattening it in JS would work equally well and put the shape in two places.
    /// </para>
    /// </summary>
    internal static FlattenedTimings Flatten(LyricsTimingsDocument document)
    {
        var starts = new List<long>();
        var ends = new List<long>();
        var lineOf = new List<int>();
        var wordOf = new List<int>();

        for (var lineIndex = 0; lineIndex < document.Lines.Count; lineIndex++)
        {
            var line = document.Lines[lineIndex];
            if (!line.IsTimed)
            {
                continue;
            }

            if (line.Words.Count == 0)
            {
                // A placed line whose words were not. It highlights as a single block, which is what
                // the markup renders for it too.
                starts.Add(line.StartMs!.Value);
                ends.Add(line.EndMs!.Value);
                lineOf.Add(lineIndex);
                wordOf.Add(0);
                continue;
            }

            for (var wordIndex = 0; wordIndex < line.Words.Count; wordIndex++)
            {
                var word = line.Words[wordIndex];
                if (!word.IsTimed)
                {
                    continue;
                }

                starts.Add(word.StartMs!.Value);
                ends.Add(word.EndMs!.Value);
                lineOf.Add(lineIndex);
                wordOf.Add(wordIndex);
            }
        }

        return new FlattenedTimings(
            starts.ToArray(),
            ends.ToArray(),
            lineOf.ToArray(),
            wordOf.ToArray(),
            document.DurationMs);
    }

    /// <summary>
    /// Mark a word as selected.
    /// </summary>
    /// <remarks>
    /// Applied as a class by the JS module rather than by re-rendering. Selection changes on every
    /// click, and re-rendering a few thousand spans to move one outline would be a diff of the whole
    /// panel each time.
    /// </remarks>
    public async Task HighlightSelectionAsync(int lineIndex, int wordIndex)
    {
        if (_module is null)
        {
            return;
        }

        try
        {
            await _module.InvokeVoidAsync("setSelectedWord", InstanceId, lineIndex, wordIndex);
        }
        catch (JSDisconnectedException)
        {
            // Circuit gone.
        }
    }

    /// <summary>
    /// Drop the selection outline.
    ///
    /// <para>
    /// Reuses the module's own setter with indices no word can have: it clears the previous outline
    /// before looking up the new one, and finds nothing, which is exactly the wanted effect without
    /// a second entry point that would have to keep in step with the first.
    /// </para>
    /// </summary>
    public Task ClearSelectionAsync() => HighlightSelectionAsync(-1, -1);

    /// <summary>Called from JS when a creator clicks a word in editor mode.</summary>
    [JSInvokable]
    public async Task WordClicked(int lineIndex, int wordIndex)
    {
        if (OnWordSelected.HasDelegate)
        {
            await OnWordSelected.InvokeAsync(new LyricsWordSelection(lineIndex, wordIndex));
        }
    }

    public async ValueTask DisposeAsync()
    {
        if (_module is not null)
        {
            try
            {
                // Cancels the animation frame loop and drops the module's listeners and element
                // cache. Without it a library page that has played a dozen cards leaves a dozen loops
                // running against detached audio elements.
                await _module.InvokeVoidAsync("dispose", InstanceId);
                await _module.DisposeAsync();
            }
            catch (JSDisconnectedException)
            {
                // The circuit is already gone, which took the module with it.
            }
            catch (Exception ex)
            {
                Logger.LogDebug(ex, "Lyrics scroller disposal was not clean.");
            }
        }

        _selfRef?.Dispose();
    }
}

/// <summary>Which word a creator clicked.</summary>
public readonly record struct LyricsWordSelection(int LineIndex, int WordIndex);

/// <summary>Every timed word as parallel arrays, ready for a per-frame lookup.</summary>
public sealed record FlattenedTimings(
    long[] Starts,
    long[] Ends,
    int[] LineOf,
    int[] WordOf,
    long DurationMs);
